### ISSUE TITLE
Inference on dataset instead of pre-computed tensor

### ISSUE THREAD

**Issue by vltanh** [KevinMusgrave/pytorch-metric-learning#296]
The current InferenceModel requires us to create a tensor of the whole gallery images, which may be infeasible in cases of large gallery. Another excess is the `emb_dim` argument since it can be inferred from the output embedding.

```python
train_vectors = torch.stack([dataset[i][0] for i in range(len(dataset))], dim=0)
inference_model.train_indexer(train_vectors, emb_dim = 64)
```

Checking the code, it is simply

```python
    def train_indexer(self, tensors, emb_dim):
        if isinstance(tensors, list):
            tensors = torch.stack(tensors)

        embeddings = torch.Tensor(len(tensors), emb_dim)
        for i in range(0, len(tensors), self.batch_size):
            embeddings[i : i + self.batch_size] = self.get_embeddings(
                tensors[i : i + self.batch_size], None
            )[0]

        self.indexer.train_index(embeddings.cpu().numpy())
```

In my opinion, another way is to do extraction directly from the dataset, and remove the need to pass the `emb_dim` this way:

```python
    def train_indexer(self, dataset):
        dl = DataLoader(dataset, batch_size=self.batch_size)
        embeddings = []
        for inp in dl:
            embeddings.append(self.get_embeddings(inp, None)[0])
        embeddings = torch.cat(embeddings)
        self.indexer.train_index(embeddings.cpu().numpy())
```

Additionally, the `get_nearest_neighbors` only returns the indices. The current way is to get the images directly from `train_vectors` by slicing. But we can also do this on the dataset:

```python
def get_item(dataset, indices):
    d = []
    for i in indices:
        d.append(dataset[i].unsqueeze(0))
    return torch.cat(d)
```

I think it is better in terms of space efficiency although it implies some restrictions on the way the dataset is implemented (`get_item` should only return the input).

**Comment by KevinMusgrave** [comment]
You're right, passing in a dataset would make more sense.

If you've already made the changes, feel free to make a pull request. (It doesn't have to be perfect)
